### PR TITLE
feat: document_loader Textract linearized output

### LINKED ISSUE #12068: feat: Add Linearized output to Textract PDFLoader
### Feature request

Textract released the [LAYOUT](https://docs.aws.amazon.com/textract/latest/dg/layoutresponse.html) feature, which identifies different layout elements like tables, lists, figures, text-paragraphs and titles. This should be used by the AmazonTextractPDFParser to generate a linearized output to improve downstream LLMs accuracy with those hints.

Text output should render tables and key/value pairs and text in reading order for multi-column text and prefix lists with a *, when features like LAYOUT, TABLES, FORMS are passed to the textract call

### Motivation

Improve downstream LLM accuracy

### Your contribution

I'll submit a PR for this feature.

### PR DESCRIPTION
**Pull request by schadem**
<!-- Thank you for contributing to LangChain!

Replace this entire comment with:
  - **Description:** linearizing the output from Amazon Textract into a layout with tables and key/value pairs when those features are passed in as feature parameters 
  - **Issue:** fixes #12068
  - **Dependencies:** adding amazon-textract-textractor as dependency
  - **Tag maintainer:** @3coins  
  - **Twitter handle:**

### CODE REVIEW

**Comment by baskaryan** [comment]
cc @3coins

File: libs/langchain/langchain/document_loaders/parsers/pdf.py [hunk 1]
```diff
@@ -376,6 +376,48 @@ class AmazonTextractPDFParser(BaseBlobParser):
...
+        "example_data/alejandro_rosalez_sample-small.jpeg",
+        textract_features=["TABLES", "LAYOUT"])
+    documents = loader.load()
+    ```
+
+    it will generate output that formats the text in reading order and
+    try to output the information in a tabular structure or
+    output the key/value pairs with a colon (key: value).
+    This helps most LLMs to achieve better accuracy when
+    processing these texts.
```
Comment (3coins) [inline libs/langchain/langchain/document_loaders/parsers/pdf.py:419]:
```suggestion
    AmazonTextractPDFLoader also supports linearization of the output, which maintains the structure of the input 
    document. For example, when using LAYOUT feature, the reading order of the text is maintained. Similarly, 
    when selecting FORMS feature, information is extracted into key-value pairs (key: value). With TABLES feature, 
    any tables present in the document are extracted into a tabular structure. These features helps the LLMs 
    process the information with better accuracy. Here is an example of using the TABLES and LAYOUT feature.
    
    ```python
    from langchain.document_loaders import AmazonTextractPDFLoader
    # you can mix and match each of the features
    loader=AmazonTextractPDFLoader(
        "example_data/alejandro_rosalez_sample-small.jpeg",
        textract_features=["TABLES", "LAYOUT"])
    documents = loader.load()
    ```
```

File: libs/langchain/langchain/document_loaders/parsers/pdf.py [hunk 2]
```diff
@@ -376,6 +376,48 @@ class AmazonTextractPDFParser(BaseBlobParser):
     """Send `PDF` files to `Amazon Textract` and parse them.
 
     For parsing multi-page PDFs, they have to reside on S3.
+
+    The AmazonTextractPDFLoader calls the
+    [Amazon Textract Service](https://aws.amazon.com/textract/)
+    to convert PDFs into a Document structure.
+    The loader does pure OCR at the moment, with more features like
+    layout support planned, depending on demand.
```
Comment (3coins) [inline libs/langchain/langchain/document_loaders/parsers/pdf.py:384]: Are we providing layout support here, so this should be corrected.

File: libs/langchain/tests/integration_tests/document_loaders/test_pdf.py [hunk 3]
```diff
@@ -150,14 +150,14 @@ def test_mathpix_loader() -> None:
...
-            ["FORMS", "TABLES"],
+            ["FORMS", "TABLES", "LAYOUT"],
             1,
             False,
         ),
         (str(Path(__file__).parent.parent / "examples/hello.pdf"), ["FORMS"], 1, False),
         (
             "s3://amazon-textract-public-content/langchain/layout-parser-paper.pdf",
-            None,
+            ["FORMS", "TABLES", "LAYOUT"],
```
Comment (3coins) [inline libs/langchain/tests/integration_tests/document_loaders/test_pdf.py:160]: Should we keep a test without these features?

**Review by 3coins — CHANGES_REQUESTED** [review]
@schadem 
Thanks for working on this change. 

Can you please add a small para to the PR description to summarize the changes.

I have requested some doc updates that simplify the instructions, but feel free to correct me if there is in-accurate info. Will also run through the tests and the notebook to test these changes.

File: libs/langchain/tests/integration_tests/document_loaders/test_pdf.py (hunk 3 above)
Comment (schadem) [reply libs/langchain/tests/integration_tests/document_loaders/test_pdf.py:160]: added one without FORMS and LAYOUT and one without any features

File: libs/langchain/langchain/document_loaders/parsers/pdf.py (hunk 2 above)
Comment (schadem) [reply libs/langchain/langchain/document_loaders/parsers/pdf.py:384]: fixed

[the remark this answers is not in the document]
Comment (schadem) [reply libs/langchain/langchain/document_loaders/parsers/pdf.py:384]: fixed

**Comment by anjanvb** [comment]
@3coins @baskaryan anything else needed for the merge?

**Comment by 3coins** [comment]
@schadem @anjanvb 
I don't see some of requested doc updates, this also needs a rebase from the master branch. Let me know if I missed any updates.

**Comment by anjanvb** [comment]
@3coins ah, I see now that there's a need to rebase from upstream. Also, can you point me to the doc updates review comments please?

**Comment by 3coins** [comment]
> ah, I see now that there's a need to rebase from upstream

There are conflicts that should be resolved here, so you need a rebase. 
Screenshot 2023-10-26 at 12 06 38 PM

File: libs/langchain/langchain/document_loaders/parsers/pdf.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (3coins) [reply libs/langchain/langchain/document_loaders/parsers/pdf.py:384]: @anjanvb 
This text is still talking about "more features like layout support planned, depending on demand." that doesn't make sense. Can you correct this.

File: libs/langchain/tests/integration_tests/document_loaders/test_pdf.py (hunk 3 above)
Comment (3coins) [reply libs/langchain/tests/integration_tests/document_loaders/test_pdf.py:170]: @anjanvb 
This still doesn't have a test without any features.

Comment (schadem) [reply libs/langchain/tests/integration_tests/document_loaders/test_pdf.py:170]: @3coins : (str(Path(__file__).parent.parent / "examples/hello.pdf"), [], 1, False), has not features

**Comment by schadem** [comment]
Updated the docs (again). must have sneaked back in while I did the reformatting.

**Pull request closed without merging**